### PR TITLE
Chore/spectacle mdx loader addition

### LINKED ISSUE #1164: Infra: Bring `spectacle-mdx-loader` in to the monorepo
- [ ] Add `packages/spectacle-mdx-loader`
- [ ] Add `examples/mdx` that uses the loader
- [ ] Remove old repository
- [ ] Verify we can correctly publish.

### PR DESCRIPTION
**Pull request by paulmarsicloud**
### Description

First pass and attempt at bringing in `spectacle-mdx-loader` into the monorepo. Took onboard @ryan-roemer's advice to remove all devDeps and scripts and focus on getting `lint` and `prettier` working as a start. Moved `prop-types` and `spectacle-mdx-loader` into top-level `package.json` so that `spectacle-mdx-loader` could be called directly from `examples/mdx/webpack.config.js`.

In a good spot now for a second pair of eyes to take a look through 👍 

Fixes #1164 

#### Type of Change

- [x] New feature (non-breaking change which adds functionality)
- [ ] This change requires a documentation update

### Checklist: (Feel free to delete this section upon completion)

- [ ] I have included a [changeset](../CONTRIBUTING.md#changesets) if this change will require a version change to one of the packages.
- [x] I have performed a self-review of my own code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [x] I have run `pnpm run check:ci` and all checks pass
- [ ] I have added tests that prove my fix is effective or that my feature works
- [x] My changes generate no new warnings
- [ ] Any dependent changes have been merged and published in downstream modules

### CODE REVIEW

**Review by fritz-c — CHANGES_REQUESTED** [review]

File: packages/spectacle-mdx-loader/LICENSE.txt [hunk 1]
```diff
@@ -0,0 +1,21 @@
+The MIT License (MIT)
```
[the remark this answers is not in the document]
Comment (ryan-roemer) [reply packages/spectacle-mdx-loader/LICENSE.txt:1]: Ooh, I noticed we don't ship `LICENSE.txt` in `create-spectacle` and `spectacle`. We should keep this here and _add_ them to the other packages we ship.

cc @gksander @carlos-kelly

File: packages/spectacle-mdx-loader/CONTRIBUTING.md [hunk 2]
```diff
@@ -0,0 +1,181 @@
+# Contributing
+
+Thanks for contributing!
+
+## Development
+
+### Installing dependencies
+
+We use [`yarn`](https://yarnpkg.com/en/docs/getting-started).
```
[the remark this answers is not in the document]
Comment (ryan-roemer) [reply packages/spectacle-mdx-loader/CONTRIBUTING.md:9]: Yeah, we should remove this file and integrate with root `CONTRIBUTING.md` guide.

File: examples/mdx/webpack.config.js [hunk 3]
```diff
@@ -0,0 +1,63 @@
...
+
+// Customized babel loader with the minimum we need to get `mdx` libraries
+// working, which unfortunately codegen JSX instead of JS.
+const babelLoader = {
+  loader: require.resolve('babel-loader'),
+  options: {
+    // Use user-provided .babelrc
+    babelrc: true,
+    // ... with some additional needed options.
+    presets: [require.resolve('@babel/preset-react')]
```
[the remark this answers is not in the document]
Comment (ryan-roemer) [reply examples/mdx/webpack.config.js:12]: Hmmm -- not sure here. The root babel config has TS and stuff we don't need. For here we only need enough to support the demo. Could try just removing this field and see if it still works.

File: package.json [hunk 4]
```diff
@@ -33,10 +33,12 @@
     "mkdirp": "^1.0.4",
     "nps": "^5.10.0",
     "prettier": "^2.4.1",
+    "prop-types": "^15.7.2",
     "puppeteer": "^16.0.0",
     "raw-loader": "^4.0.0",
     "rimraf": "^3.0.0",
     "serve": "^14.0.1",
+    "spectacle-mdx-loader": "workspace:*",
```
Comment (ryan-roemer) [inline package.json:41]: These deps should be in `examples/mdx/package.json`

File: package.json [hunk 5]
```diff
@@ -171,6 +174,7 @@
         "./website:lint",
```
Comment (ryan-roemer) [inline package.json:174]: GLOBAL: We should add all the `"./packages/spectacle-mdx-loader:*"` here on all the applicable wireit tasks.

File: examples/mdx/index.js [hunk 6]
```diff
@@ -0,0 +1,60 @@
...
+    <Box padding="1em">
+      <Progress />
+    </Box>
+  </FlexBox>
+);
+// SPECTACLE_CLI_TEMPLATE_END
+
+const Presentation = () => (
+  <MDXProvider components={mdxComponentMap}>
+    <Deck loop theme={theme} template={template}>
```
[the remark this answers is not in the document]
Comment (paulmarsicloud) [reply examples/mdx/index.js:45]: That I am going to have to refer to the experts on (@ryan-roemer / @carlos-kelly ) 😄

File: examples/mdx/slides.mdx [hunk 7]
```diff
@@ -0,0 +1,66 @@
...
+export const testProp = 50;
+
+# Spectacle Presentation (MDX) 👋
+
+This example uses:
+
+- `mdx`
+- `@babel/preset-react` for the React components we import
+
+Notes: These are presenter notes. This text will be automatically exported as `notes` from either your MDX file (e.g. `import {notes} from './my-file.mdx'`).
```
[the remark this answers is not in the document]
Comment (paulmarsicloud) [reply examples/mdx/slides.mdx]: updated!

File: examples/mdx/test-component.js [hunk 8]
```diff
@@ -0,0 +1,12 @@
+import React from 'react';
+import PropTypes from 'prop-types';
```
[the remark this answers is not in the document]
Comment (paulmarsicloud) [reply examples/mdx/test-component.js:2]: this guy? https://github.com/FormidableLabs/spectacle/blob/chore/spectacle-mdx-loader-addition/examples/mdx/package.json#L11

File: examples/mdx/slides.mdx [hunk 9]
```diff
@@ -0,0 +1,66 @@
...
+
+Standard Text
+
+---
+
+> Example Quote
+
+---
+
+```javascript
```
[the remark this answers is not in the document]
Comment (paulmarsicloud) [reply examples/mdx/slides.mdx]: I'm not familiar either, but I changed it to `jsx` and it worked!

File: examples/mdx/webpack.config.js (hunk 3 above)
[the remark this answers is not in the document]
Comment (paulmarsicloud) [reply examples/mdx/webpack.config.js]: I've removed the option altogether and it still ran fine to support the demo. If you would like me to add certain things back, I'm happy to do so!

File: packages/spectacle-mdx-loader/CONTRIBUTING.md (hunk 2 above)
[the remark this answers is not in the document]
Comment (paulmarsicloud) [reply packages/spectacle-mdx-loader/CONTRIBUTING.md]: Removed!

File: packages/spectacle-mdx-loader/README.md [hunk 10]
```diff
@@ -0,0 +1,82 @@
...
+## Install
+
+```sh
+$ npm add --save-dev spectacle-mdx-loader
+$ yarn add --dev spectacle-mdx-loader
+```
+
+## Usage
+
+To use this loader in a Spectacle presenation you need to configure webpack and then add the surrounding MDX Spectacle helper code. See a [full example](./examples/mdx) for more details.
```
[the remark this answers is not in the document]
Comment (paulmarsicloud) [reply packages/spectacle-mdx-loader/README.md]: updated!

File: package.json (hunk 4 above)
Comment (paulmarsicloud) [reply package.json]: these have been moved 👍

File: package.json (hunk 5 above)
Comment (paulmarsicloud) [reply package.json:173]: I've updated these, let me know if I've missed anything!

File: packages/spectacle-mdx-loader/src/helpers.js [hunk 11]
```diff
@@ -0,0 +1,112 @@
...
+/*
+ * See this link for an explanation of the regex solution:
+ * https://stackoverflow.com/questions/6462578/regex-to-match-all-instances-not-inside-quotes/23667311#23667311
+ * Note that the regex isn't concerned about code blocks (```).
+ * Tracking pairs of ` should be sufficient to capture code blocks, too.
+ */
+const MOD_REG = /\\`|`(?:\\`|[^`])*`|(^(?:import|export).*$)/gm;
+
+const NOTES_MARKER = 'Notes: ';
+const NOTES_REG = new RegExp(`^${NOTES_MARKER}`, 'm');
```
[the remark this answers is not in the document]
Comment (paulmarsicloud) [reply packages/spectacle-mdx-loader/src/helpers.js:15]: this too I'm going to have to defer to the expert (@ryan-roemer) 😄

[the remark this answers is not in the document]
Comment (ryan-roemer) [reply packages/spectacle-mdx-loader/src/helpers.js:15]: Let's just add a ticket to add tests to `spectacle-mdx-loader` as we have none.

File: examples/mdx/index.js (hunk 6 above)
[the remark this answers is not in the document]
Comment (ryan-roemer) [reply examples/mdx/index.js:45]: Good find @fritz-c 

@paulmarsicloud it seems to be a long gone option. Maybe remove `loop` and re-check everything works?

File: examples/mdx/test-component.js (hunk 8 above)
[the remark this answers is not in the document]
Comment (fritz-c) [reply examples/mdx/test-component.js:2]: Oh, ok. 👍

File: examples/mdx/package.json [hunk 12]
```diff
@@ -0,0 +1,71 @@
...
+    "name": "spectacle-example-mdx",
+    "private": true,
+    "dependencies": {
+      "spectacle": "*",
+      "react": "^18.1.0",
+      "react-dom": "^18.1.0"
+    },
+    "devDependencies": {
+      "@mdx-js/react": "^1.5.3",
+      "prop-types": "^15.7.2",
```
Comment (ryan-roemer) [inline examples/mdx/package.json:11]: I think this should be a prod dep?

File: examples/mdx/package.json [hunk 13]
```diff
@@ -0,0 +1,71 @@
+{
+    "name": "spectacle-example-mdx",
+    "private": true,
+    "dependencies": {
+      "spectacle": "*",
+      "react": "^18.1.0",
+      "react-dom": "^18.1.0"
+    },
+    "devDependencies": {
+      "@mdx-js/react": "^1.5.3",
```
Comment (ryan-roemer) [inline examples/mdx/package.json:10]: I think this one too should be a prod dep?

File: examples/mdx/webpack.config.js [hunk 14]
```diff
@@ -0,0 +1,56 @@
...
+    // within _this_ project's `node_modules` traversal tree.
+    rules: [
+      {
+        test: /\.jsx?$/,
+        use: [babelLoader]
+      },
+      // `.md` files are processed as pure text.
+      {
+        test: /\.md$/,
+        use: [require.resolve('raw-loader')]
```
Comment (ryan-roemer) [inline examples/mdx/webpack.config.js:36]: We need a `devDependencies` for this.

File: examples/mdx/webpack.config.js [hunk 15]
```diff
@@ -0,0 +1,56 @@
+const path = require('path');
+const HtmlWebpackPlugin = require('html-webpack-plugin');
+const base = require('../../webpack.config.base');
+
+// Customized babel loader with the minimum we need to get `mdx` libraries
+// working, which unfortunately codegen JSX instead of JS.
+const babelLoader = {
+  loader: require.resolve('babel-loader')
```
Comment (ryan-roemer) [inline examples/mdx/webpack.config.js:8]: We need a `devDependencies` for this (technically in root, but should have here if we're not getting config from root).

File: examples/mdx/index.js (hunk 6 above)
[the remark this answers is not in the document]
Comment (ryan-roemer) [reply examples/mdx/index.js]: I removed this.

**Comment by ryan-roemer** [comment]
@paulmarsicloud I've finished my cleanup of examples configs and dependencies. Mind double checking my commits?

@carlos-kelly @gksander @fritz-c I've noticed a discrepancy in code blocks in MD vs. MDX.

For our MD example with `pnpm start:md` for the code page we have:

Screen Shot 2022-09-21 at 1 50 03 PM

For our MDX example with `pnpm start:mdx` for the code page with exact same sample we have:

Screen Shot 2022-09-21 at 1 50 08 PM

Any ideas here? Thanks!

**Comment by ryan-roemer** [comment]
Offhand, now that I'm looking at things looks like our `Markdown` component does a **lot** more than `mdxComponentMap` does alone -- and we recommend in our guide to just use the `mdxComponentMap` for MDX decks.

**Comment by ryan-roemer** [comment]
Per slack discussion, we're not blocking on this and I've filed https://github.com/FormidableLabs/spectacle/issues/1232 to capture.

**Review by ryan-roemer — APPROVED** [review]
Great work Paul! 🎉

**Comment by paulmarsicloud** [comment]
> @paulmarsicloud I've finished my cleanup of examples configs and dependencies. Mind double checking my commits?

Commits look good and works proper on my end!

File: packages/spectacle-mdx-loader/src/helpers.js (hunk 11 above)
[the remark this answers is not in the document]
Comment (ryan-roemer) [reply packages/spectacle-mdx-loader/src/helpers.js:15]: Added https://github.com/FormidableLabs/spectacle/issues/1235.

**Comment by ryan-roemer** [comment]
@paulmarsicloud We're missing a `build` task for `examples/mdx` which I'm going to add (and to CI)

**Comment by ryan-roemer** [comment]
I've added the `build` task. I think we're otherwise ready for this PR with all the rest of the follow-on work ticketed.
